src/dbnode/encoding/proto/iterator.go

```diff
@@ -0,0 +1,878 @@
+	bitsetLengthBits, err := it.readVarInt()
+	if err != nil {
+		return err
+	}
+
+	for i := uint64(0); i < bitsetLengthBits; i++ {
+		bit, err := it.stream.ReadBit()
+		if err != nil {
+			return fmt.Errorf("%s error reading bitset: %v", itErrPrefix, err)
+		}
+
+		if bit == opCodeBitsetValueIsSet {
+			// Add 1 because protobuf fields are 1-indexed not 0-indexed.
+			it.bitsetValues = append(it.bitsetValues, int(i)+1)
+		}
+	}
+
+	return nil
+}
+
+func (it *iterator) readVarInt() (uint64, error) {
+	var (
+		// Convert array to slice and reset size to zero so
+		// we can reuse the buffer.
+		buf      = it.varIntBuf[:0]
+		numBytes = 0
+	)
+	for {
+		b, err := it.stream.ReadByte()
+		if err != nil {
+			return 0, fmt.Errorf("%s error reading var int: %v", itErrPrefix, err)
+		}
+
+		buf = append(buf, b)
+		numBytes++
+
+		if b>>7 == 0 {
+			break
+		}
+	}
+
+	buf = buf[:numBytes]
+	varInt, _ := binary.Uvarint(buf)
+	return varInt, nil
+}
+
+func (it *iterator) readIntSig(i int) error {
+	updateControlBit, err := it.stream.ReadBit()
+	if err != nil {
+		return fmt.Errorf(
+			"%s error reading int significant digits update control bit: %v",
+			itErrPrefix, err)
+	}
+	if updateControlBit == opCodeNoChange {
+		// No change.
+		return nil
+	}
+
+	sigDigitsControlBit, err := it.stream.ReadBit()
+	if err != nil {
+		return fmt.Errorf(
+			"%s error reading zero significant digits control bit: %v",
+			itErrPrefix, err)
+	}
+	if sigDigitsControlBit == m3tsz.OpcodeZeroSig {
+		it.customFields[i].intSigBitsTracker.NumSig = 0
+	} else {
+		numSigBits, err := it.readBits(6)
+		if err != nil {
+			return fmt.Errorf(
+				"%s error reading number of significant digits: %v",
+				itErrPrefix, err)
+		}
+
+		it.customFields[i].intSigBitsTracker.NumSig = uint8(numSigBits) + 1
+	}
+
+	return nil
+}
+
+func (it *iterator) readIntValDiff(i int) error {
+	negativeControlBit, err := it.stream.ReadBit()
+	if err != nil {
+		return fmt.Errorf(
+			"%s error reading negative control bit: %v",
+			itErrPrefix, err)
+	}
+
+	numSig := int(it.customFields[i].intSigBitsTracker.NumSig)
+	diffSigBits, err := it.readBits(numSig)
+	if err != nil {
+		return fmt.Errorf(
+			"%s error reading significant digits: %v",
+			itErrPrefix, err)
+	}
+
+	if it.customFields[i].fieldType == cUnsignedInt64 {
+		diff := uint64(diffSigBits)
+		shouldSubtract := false
+		if negativeControlBit == opCodeIntDeltaNegative {
+			shouldSubtract = true
+		}
+
+		prev := uint64(it.customFields[i].prevFloatBits)
+		if shouldSubtract {
+			it.customFields[i].prevFloatBits = prev - diff
+		} else {
+			it.customFields[i].prevFloatBits = prev + diff
+		}
+	} else {
+		diff := int64(diffSigBits)
+		sign := int64(1)
+		if negativeControlBit == opCodeIntDeltaNegative {
+			sign = -1.0
+		}
+
+		prev := int64(it.customFields[i].prevFloatBits)
+		it.customFields[i].prevFloatBits = uint64(prev + (sign * diff))
+	}
+
+	return nil
+}
+
+// TODO: Share logic with encoder if possible
```

Comment:
+1, logic looks identical.

Reply:
yeah the problem is they operate on different types so I would have to do some `interface{}` magic which is gonna allocate :/